### ISSUE TITLE
overloaded CPU multiple unrelated files analyzed - only occurring in M1 mac compared against Linux

### ISSUE THREAD

**Issue by alexlatif** [microsoft/pylance-release#4399]
One file change, no extensions but Pylance, in particular on deletion cause immediately over 20 unrelated files to be analyzed and I feel like I have followed all VScode workspace setting to configure language server. Currently the only fix is to use VScode without Pylance

Pylance v2023.5.30 
VScode Version: 1.78.2
Python 3.11

using "files.exclude", "files.watcherExclude" and "search.exclude" with all these files

            "**/webpack-stats.json": true,
            "**/webpack-bundle-debug-stats.json": true,
            "**/venv/**": true,
            "**/venv_old": true,
            "**/staticfiles": true,
            "**/package-lock*": true,
            "**/dist": true,
            "**/dev_media": true,
            "**/local_junk": true,
            "**/coverage": true,
            "**/htmlcov": true,
            "**/__pycache__": true,
            "**/rust/**/target/**": true,
            "**/.vscode/**": true,
            "**/.mypy_cache/**": true,
            "**/.git/**": true,
            
it seems it is still trying to analyze them. every change especially deletion causes over 20 files to be analyzed and my CPU just gets eaten, code freezes and I am stuck but to remove Pylance altogether. I have no other extensions running.

Could you please help me debug what is causing this and if you need anything else to debug further.

[pylance_logs.txt](https://github.com/microsoft/pylance-release/files/11524753/pylance_logs.txt)

**Comment by nitishxp** [comment]
I am also facing the same issue in the Windows system, pylance seems to be analyzing a whole bunch of unopen files and causing high CPU usage which forced me to remove the extension.

**Comment by debonte** [comment]
@alexlatif:

1. What do you have `python.analysis.diagnosticMode` set to? `openFilesOnly` or `workspace`? I believe it's expected that upon deleting a file, those files (i.e. your open files or your entire workspace) would be re-analyzed. @heejaechang, is that correct?
2. Can you explain what you did in the logged session? What file did you delete and which files were unexpectedly analyzed?

**Comment by heejaechang** [comment]
ya, currently, if a file structure is changed (a file is added or removed), then it can affect how `import statement` is resolved, so we dump all caches and re-create all information from scratch. and that is `by design`.

by the way, according to the log, most of time is used by `indexing workspace files`, does it still happen with the latest `prerelease pylance` (`2023.5.31`)? we recently changed what files to `reindex` when workspace is `invalidated`

if it is still happening and if you don't see benefit of `indexing` due to its `perf impact`, you can turn it off using `python.analysis.indexing: false` in `settings.json`

**Comment by alexlatif** [comment]
@heejaechang i am not changing the file structure merely a few lines of within 1 file.

**Comment by alexlatif** [comment]
@debonte I rolled back to VScode 1.77 - same issue - and then i rolled back to 1.75 - fixed. Analyzing 1 file for a mere deletion of a few lines. Will prevent auto update and use this version until the latest release is fixed. Please let me know how i can help.

**Comment by debonte** [comment]
> then i rolled back to 1.75 - fixed.

Are you still using Pylance 2023.5.30 on vscode 1.75? Looks like 2023.5.30 should still be supported, but just wanted to double-check.

I'm not sure why rolling back VS Code would have an impact. Maybe their file watcher implementation changed.

> merely delete a couple of lines inside the file `/Users/alessandrolatif/watz/backend/apps/metric_handler/metrics/activity/cycling_dynamics_activity_metric.py`
>
> this then prompts another couple of hundred files

When you make edits to a file, it appears that we also reanalyze any files (open files with `openFilesOnly` or all files with `workspace`) that import that changed file, and files that import those files, etc.

Do the files you see getting analyzed after changing `cycling_dynamics_activity_metric.py` import `cycling_dynamics_activity_metric.py` or import files that import `cycling_dynamics_activity_metric.py`?
 
Btw, shortly after you change `cycling_dynamics_activity_metric.py` this `auto_generated` file is changed:  `/Users/alessandrolatif/watz/backend/auto_generated/__init__.py`. Wondering if that may be making the situation worse.

**Comment by heejaechang** [comment]
@alexlatif if you are talking about `vscode` version not `pylance` version, then it must be `NOT` related to our extension (`pylance`)

if it is `pylance` issue, `vsocde` shouldn't matter. 

according to your post, this is version you were using,

```
Pylance v2023.5.30
VScode Version: 1.78.2
Python 3.11
```

just to make sure it is related to us, can you move back to vscode `1.78.2`, and just rollback `pylance` extension to see whether that resolve the issue? and let us know which pylance version resolved the issue so that we can see what kind of changes are added after that?

that said, 
> I just set it to "openFilesOnly" and I get no logs on the Python Language server.

`log` is not related to `diagnosticMode` but related to `python.analysis.logLevel`. can you set the `logLevel` to `Trace` and see whether you got the log?

> this then prompts another couple of hundred files. When i run mode "openFilesOnly" it shows the usual 20-30 files analyzed but i cant seem to get output logs.

First, the difference between `openFilesOnly` and `workspace` is that, `openFilesOnly` will only check errors for files that are currently opened in the `editors`. but `workspace` will check files that are currently closed (not opened in editors) as well. and according to your log, you have `Found 666 source files`, so initially we will analyze all those `666` files if the mode is set to `workspace` and that could take quite a bit of time depends on context and code.

and now, when a file is changed 

> merely delete a couple of lines inside the file

the amount of text you changed doesn't matter. the issue is whether the change can cause errors in other files or not. so, we need to re-analyze all files that could depends on the file you just changed transitively.

For example, you might delete `class A:...` from `a.py` and there is `b.py` which has `from a import *` and there is `c.py` which has `from b import *; o = A()`, then we need to show errors on `c.py` since `A` no longer exist, and no error on `b.py` or you could rename class `A` to class `B` and b.py happen to use `B` in them, then that error go away but new error show up in `c.py` and etc

so, I think whatever file you changed, I think there are 20~30 files that are transitively depends on the file you changed. and that is why we re-analzye them.

>  usual 20-30 files analyzed

that said, if only 1 file is opened in the editor among that 20-30 files and mode is set to `openFilesOnly`, then only that file will be re-analyzed. if you have set `workspace`, then all 20-30 files will be reanalyzed.

if you think, it doesn't work that way, we will probably need repro steps (sample code that shows the issue or github repo or etc) to dig into more.

**Comment by alexlatif** [comment]
Ok so after more tests rolling Pylance did not change much, what does definitively sort it is rolling back VScode (with latest Pylance v2023.5.30).

What is strange is that its not that the file number changes as I now understand its looking for dependancies across these files. They are often being imported into that specific file and not the other way around. It is actually the way it analyzes these files, on a newer VScode version it thinks it finds one maybe two files, but then it realizes the dependency is larger and so the file amount grows and grows before reducing. This seems to cause the higher usage.

When i roll back VScode the file amount is immediately found and resolved quickly without fuss.

I understand on the auto_generate its internal and should not cause too much craziness for now but something to keep and eye on - thanks for the reminder.

In summary I dont know how to proceed other than remain on VScode 1.75.1

**Comment by heejaechang** [comment]
> hey are often being imported into that specific file and not the other way around. It is actually the way it analyzes these files, on a newer VScode version it thinks it finds one maybe two files, but then it realizes the dependency is larger and so the file amount grows and grows before reducing. This seems to cause the higher usage.

I am not sure whether I am following.

there are 2 kinds of dependencies. one is "imported by" and the other is "importing". if there are `a.py` and `b.py`, and `a.py` has `from b import *`, then `a.py` is `importing` `b.py` and `b.py` is `imported by` `a.py`

For your case, we only care `imported by` dependency. so, any file that transitively importing your file should be discovered and re-analyzed.

by the way, unless you make a change and wait until `idle`, it will be hard to track how many files are actually affected since we will `accumulate` all files we need to recheck for all your changes as you make your changes. if previous change require 10 files to be checked and you made another change before that 10 files are re-analyzed, and new change caused another 15 files to be rechecked and among them 8 are shared, then you will get (17 - whatever files already processed) files left to be processed and so on.

that said, this should be deterministic and vscode version should `not` affect this behavior (unless there is some other extension that randomly change python file behind the scene like formatter automatically updating py file when a file is saved or the `auto_generate` stuff you mentioned)

if you feel like number of files we re-analyze is not deterministic, it would be helpful for us to track down what went wrong if you can provide us some repo where we can repro the issue.
